### PR TITLE
Geopy 1451

### PR DESCRIPTION
**Pull request by domfournier**

### CODE REVIEW

File: docs/case_study/mesh_design.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# <p style="page-break-after:always;"></p>
+#
+# # Mesh design
+#
+# This section provides details about the mesh creation for the individual inversions using the `Octree Mesh Creation` routine from [Mirageoscience\geoapps](https://mirageoscience-octree-creation-app.readthedocs-hosted.com/en/latest) package. The goal is to create inversion meshes that are well adapted to each survey, while also optimal for the subsequent joint inversion process.
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply docs/case_study/mesh_design.py:19]: .. Or I guess we could just point to the octree-creation-app docs for single mesh parameters, and just focus on the joint details here.  Then you'd just have to have a sentence at the top that says: "For individual inversions, the mesh creation process is already well documented in _link_.  Therefore we will restrict the discussion here to details regarding joint mesh blah blah blah."

File: docs/case_study/unconstrained_inversion.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#
+#
+# ```{figure} ./images/grav_gradio_model_TOP.png
+# ---
+# height: 400px
+# name: grav_gradio_model_TOP
+# ---
+# (Top) (left) Observed and (right) predicted gxx data after convergence of the L2 (smooth) solution. Most of the data is well recovered.
+#
+# (Bottom) Horizontal slice for (left) the smooth and (right) L1-norm density contrast models. The outline of the inferred ore shell is shown in gray for reference.
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply docs/case_study/unconstrained_inversion.py:78]: reference iterations numbers in caption

File: docs/case_study/unconstrained_inversion.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+# ```{figure} ./images/dc_model_TOP.png
+# ---
+# height: 400px
+# name: dc_model_TOP.png
+# ---
+# (Top) (left) Observed and (right) predicted apparent resistivity data after convergence.
+#
+# (Bottom)
+# (Left) Horizontal section at 240 m elevation through the compact density model. Outline of the hematite shell is shown in grey for reference.
+# (Right) Vertical NS section through the recovered smooth resistivity model.
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply docs/case_study/unconstrained_inversion.py:103]: no shell

**Review by benk-mira — CHANGES_REQUESTED** [review]
Wow, that was a whopper!  Looks really great overall!  I caught some spelling mistakes, maybe try to run a spellchecker on the .py files? Also caught several errors in the figure captions, so maybe have a closer look when doing round 2.

I also have a couple of general comments:
1. More images of ui.  Break into ui groups and try to reference more regularily in the text.
2. I think the tutorials would have been better written up in terms of an individual inversion first, and then a joint, rather than mashing the topics together.

File: docs/inversion/data_misfit.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+#
+# $$
+# \phi_d =(F(\mathbf{m}) - \mathbf{d}^{obs})^T \mathbf{W}_d^T \mathbf{W}_d (F(\mathbf{m}) - \mathbf{d}^{obs})\;,
+# $$
+#
+# where $F(\mathbf{m})$ is a forward simulation of data (gravity, magnetic, EM, etc.) due to a model $m$. The $\mathbf{W}_d$ is a diagonal matrix holding the data weights ($\frac{1}{\mathbf{w}}$), also referred to as **data uncertainties**. The larger the weights, the less importance a given datum has on the solution. There are many sources of **noise** that can yield large data residuals:
+#
+#
+#  - Experimental noise from $\mathbf{d}^{obs}$, generally introduced during acquisition (positioning, instrumental noise)
+#  - Numerically noise introduced by $F(\mathbf{m})$ from our ability to accurately simulate the physics (discretization, interpolation, etc.)
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/data_misfit.py:35]: Done

File: docs/inversion/data_misfit.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+#
+# # Data Misfit
+#
+# This section focuses on the misfit function $\phi_d$ and related data uncertainties. The data misfit function measures the quality of a model $\mathbf{m}$ at **fitting the observed data**. It generally takes the form
+#
+# $$
+# \phi_d =\sum_{i=1}^{N}\left(\frac{d_i^{pred} - {d}_i^{obs}}{w_i}\right)^2 \;.
+# $$
+#
+# It measures the sum of **weighted residual** between observed ($\mathbf{d}^{obs}$) and predicted ($\mathbf{d}^{pre}$) data. In matrix form, this least-square measure can be written as
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/data_misfit.py:25]: Done

File: docs/inversion/data_misfit.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+#
+#
+#  - Experimental noise from $\mathbf{d}^{obs}$, generally introduced during acquisition (positioning, instrumental noise)
+#  - Numerically noise introduced by $F(\mathbf{m})$ from our ability to accurately simulate the physics (discretization, interpolation, etc.)
+#
+# In all cases, we may never expect the data residual to completely vanish, but we may allow it to be within an *acceptable* level of fitness for modelling purposes.
+
+# ## Data Uncertainties
+#
+# As mentioned in the [Data Misfit](#Data-Misfit) section, we want to assign **data uncertainties** that reflect our ability to replicate the observed data, within an acceptable level. While many strategies can be used, we are going to demonstrate basic principles to assess the effectiveness of a chosen method. We will employ two basic strategies:
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/data_misfit.py:41]: DOne

File: docs/inversion/regularization.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+# where $\mathbf{m}_{ref}$ is a reference model. This function tries to keep the model "small", as large deviations from the reference value are penalized more heavily. The reference model can vary in complexity, from a constant background value to a full 3D geological representation of the physical property. This constitutes our first constraint.
+#
+# ![ref_model](./images/reference_model.png)
+#
+#
+# (smooth-ref)=
+#
+# ### Model smoothness
+#
+# A second set of terms can be added to the regularization function to apply constraints on the model gradients, or roughness of the solution. Following the notation used in {cite:p}`LiOldenburg1998` and others before,
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:54]: Agreed. Done

File: docs/inversion/regularization.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+# ---
+
+# # Regularization
+#
+# This section focuses on the regularization functions, or constraints, that can be used to inject "geological knowledge" into the inversion process. More specifically, this section covers the weighted least-squares regularization functions:
+#
+# - [Conventional L2-norm](l2-norm)
+# - [General Lp-norms](lp-norm)
+#
+# While often referred to as an "unconstrained inversion", one could argue that the conventional model norm regularizations do still incorporate some degree of geological information, at least in the form of physical property distribution. Exactly what kind of a priori information is available dedicates the complexity of the constraints. For more advanced petrophysical and/or geometric constraints, see the [Joint Inversion](#joint_inversion.ipynb) section.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:22]: Done

File: docs/inversion/regularization.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+#
+# ### Model smoothness
+#
+# A second set of terms can be added to the regularization function to apply constraints on the model gradients, or roughness of the solution. Following the notation used in {cite:p}`LiOldenburg1998` and others before,
+#
+# $$
+# f_x(\mathbf{m}) = \mathbf{G}_x (\mathbf{m} - \mathbf{m}_{ref}) \;,
+# $$
+#
+# where $\mathbf{G}_x$ is a finite different operator that measures the gradient of the model $\mathbf{m}$ along one of the Cartesian directions (`x` for Easting). This function keeps the model "smooth", as large gradients (sharp contrasts) are penalized strongly. Two additional terms are needed to measure the model gradients along the Northing ($f_y$) and vertical ($f_z$) direction.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:60]: Done

File: docs/inversion/regularization.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+#
+# Note that we have added scaling parameters $\alpha_i$ to each function. The role of those multipliers is two-fold:
+#
+# - User-defined weights to emphasize the contribution of a particular function
+#
+# - Dimensionality scaling to level the functions with each other.
+#
+# **Note**
+#
+# For all SimPEG inversions, dimensionality scaling is applied directly by the program, allowing the default state to be all 1s for simplicity.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:79]: Add notes

File: docs/inversion/regularization.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+# \| f(\mathbf{m})_i \|_p \approx \| \mathbf{R}_i f(\mathbf{m})_i\|
+# $$
+#
+# where
+#
+# $$
+# \mathbf{R}_i = {\frac{1}{\left( {{f(m)^{(k-1)}_i}}^{2} + \epsilon^2 \right)^{1-p/2 }}}\,.
+# $$
+#
+# Here the superscript $(k)$ denotes the inversion iteration number. This is also unknown as an iterative re-weighted least-squares (IRLS) method. For more details on the implementation refer to {cite:p}`fournier_2019`.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:138]: Done

File: docs/inversion/regularization.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+#
+# That is, we ask the inversion to recover anomalies that have small volumes but large physical property contrasts. This is desirable if we know the geological bodies to be discrete and compact (e.g. kimberlite pip, dyke, etc.). The example below demonstrates the outcome of an inversion for different combinations of norms applied to the model
+#
+# ![sparse_model](./images/sparse_models.png)
+#
+# The [figure above](sparse_model) shows vertical sections through the true and recovered models (from left to right) with L2, L1 and L0-norm on the model values.
+#
+# Note that as $p \rightarrow 0$ the volume of the anomaly shrinks to a few non-zero elements while the physical property contrasts increase. They generally agree on the center position of the anomaly but differ greatly on the extent and shape.
+#
+# No smoothness constraints were used ($\alpha_x,y,z = 0$).
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:164]: Done

File: docs/inversion/regularization.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+#
+# ![smallness_setup](./images/smallness_norm_setup.png)
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:166]: Moved comment below image.

File: docs/inversion/regularization.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+# $$
+#
+# In this case, we are requesting a model that has either large or no (zero) gradients. This promotes anomalies with sharp contrast and constant values within.
+#
+# ![flat_model](./images/flat_models.png)
+#
+#
+# The [figure above](flat_model) shows vertical sections through the true and recovered models (from left to right) with L2, L1 and L0-norm on the model gradients. No reference model was used ($\alpha_s = 0$) with uniform norm values on all three Cartesian components.
+#
+# ![smoothness_setup](./images/smoothness_norm_setup.png)
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/regularization.py:198]: Done

File: docs/inversion/joint_inversion.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+#       jupytext_version: 1.16.1
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# # Joint Inversion
+#
+# This section introduces the methodolgy to invert multiple datasets jointly.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/joint_inversion.py:17]: Done

File: docs/inversion/joint_inversion.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+# ## Joint Surveys Inversion
+#
+#
+# As illustrated below, the `joint surveys` program allows to invert multiple datasets at once for a single physical property model. The goal is to combine complementary  information from various geophysical surveys that sense the Earth differently.
+#
+# ![joint_surveys](./images/joint_surveys.png)
+#
+# For example, a magnetotelluric survey is mostly sensitive to deep structures but can still be affected by local changes in resistivity near the sensor locations. A ground direct-current survey on the other hand is highly sensitive to near surface changes in resisivity. Inverting both surveys together would provide complementary information that improves overall our modeling capabilities.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/joint_inversion.py:26]: Done

File: docs/inversion/joint_inversion.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+# \phi_d = \sum_{j=A, B, C} \phi_d^j
+# $$
+#
+# Since each misfit tries to update the same model values, the partial derivatives of each function is also a summation, such that
+#
+# $$
+# \frac{\delta \phi_d}{\delta m} = \sum_{j=A, B, C} \frac{\delta \phi_d^j}{\delta m}
+# $$
+#
+# The joint survey inversion user requires a list of standalone inversion groups as input.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/joint_inversion.py:46]: I will deal with this when we revamp the tutorial

File: docs/inversion/joint_inversion.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+# - [Cross-gradient](cross-gradient)
+# - [Linear Correspondence](linear-correspondence)
+#
+# - [Petrophysically Guided Inversion (PGI)](pgi)
+
+# (cross-gradient)=
+#
+# ### Cross-gradient
+#
+# The `cross-gradient` regularization was first introduced by {cite:p}`gallardo2003` to constrain electrical resistivity and seismic velocity inversions, but the same strategy applies to any physical property models. As the name states, the method employes the cross product on the spatial gradients of models such that
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/joint_inversion.py:76]: Done

File: docs/inversion/joint_inversion.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+#
+# where $\nabla \mathbf{m_A}$ and $\nabla \mathbf{m_B}$ are the gradients for two distinct physical properties (density, magnetization components, resitivity, etc.). Since the cross-product of two vectors is also a vector, we use the total length (l2-norm) of the cross-product. The constraint is small (no impact) if the gradients of the models are either aligned or zero. Oppositely, the measure becomes large if edges in the physical models are perpendicular with each other. Since we attempting to minimize this function, this constraint will force model boundaries to occur at the same location or not at all.
+#
+# It is possible to constrain more than two physical property models by adding multiple cross-gradient terms for every pair of models such that
+#
+# $$
+# \phi_c(\mathbf{m_A},\mathbf{m_B}, \mathbf{m_C}) = \alpha_{AB} \phi(\mathbf{m_A},\mathbf{m_B}) + \alpha_{AC}\phi(\mathbf{m_A},\mathbf{m_C}) + \alpha_{BC} \phi(\mathbf{m_B},\mathbf{m_C})
+# $$
+#
+# and so on. Each term has a scaling parameter ($\alpha$) to control the importance of specific cross-gradient terms.
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/inversion/joint_inversion.py:90]: Yep, later

File: docs/case_study/mesh_design.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/case_study/mesh_design.py:19]: Good point

[the remark this answers is not in the document]
Comment (domfournier) [reply docs/case_study/mesh_design.py:19]: Make sense

File: docs/case_study/mesh_design.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+# name: mesh_core
+# ---
+# Core parameters used for the mesh creation of each inversion. Only the refinement strategies differ between surveys.
+# ```
+
+# ## Refinements
+#
+# An octree mesh allows to increase the mesh resolution (smaller cells) in specific regions - warranted for either numerical accuracy or modeling purposes. Fine cells can be "added" to the octree grid using various insertion methods:
+#
+# - Radial: Add concentric spherical shells of cells around points. It is used for numerical accuracy near receiver poles (DC) and dipoles (EM).
```
[the remark this answers is not in the document]
Comment (domfournier) [reply docs/case_study/mesh_design.py:92]: I will add reference to octree-docs, and push this one under the rug for now.

**Pull request merged**